### PR TITLE
Explicit dynamics Time Integrator

### PR DESCRIPTION
**Pull request by hugary1995**
This PR is based on @sveerara and @cbolisetti 's work.

This is to implement the explicit dynamics time integrator for structural dynamics.
Currently works for M u_dotdot + C u_dot + K u = F

### Tagging
Similar to the existing matrix tag "TIME", I added a new matrix tag called "SECONDTIME" to tag all matrix contributions from kernels that derive from SecondTimeDerivative. Since the SecondTimeDerivative derives from TimeKernel, it is still tagged with "TIME", therefore computing matrix associated with "TIME" will give me [M + C], and computing matrix associated with "SECONDTIME" will give me [M]. I call [M + C] the system matrix and [M] the mass matrix -- they enter the formulation naturally.

### Initial Time Derivative of the solution
Since now the system has second time derivative, we have to have proper ways of defining first time derivatives as initial conditions. I modified InitialConditionTempl to accept an optional initial time derivative of the variable. The value is optionally computed by InitialCondition::valueDot(...). I modified FunctionIC to demo this capability, the usage looks like:
```
[Variables]
  [./u]
    order = FIRST
    family = LAGRANGE

    [./InitialCondition]
      type = FunctionIC
      function = 0
      function_dot = 2*pi*x
    [../]
  [../]
[]
```

### TensorMechanics
I added an InertialForceExplicit kernel that derives from SecondTimeDerivative to demo the usage in structural dynamics. Note that the InertialForceExplicit is a temporary name before we deprecate the old InertialForce -- if we want to switch dynamically between "implicit" and "explicit" solvers, we want to split the old InertialForce so that we can tag first and second time derivatives differently.

### Tests
All existing tests for ActuallyExplicitEuler with only first time derivatives passed.
I added two tests for system that has second time derivative, and added proper initial time derivatives, they also pass.
I also added two tests in TensorMechanics for a simple 1D (but the mesh is 3D) wave propagation.

ref #9726

### CODE REVIEW

**Comment by hugary1995** [comment]
I've cleaned it up a bit.
I will be able to add C u_dot term once I figure out how to add an extra tag.

The test that has unexpected output:
tensor_mechanics/test/tests/dynamics/wave_1D/wave_explicit.i
still need to figure out why...

**Comment by hugary1995** [comment]
[MOOSEExplicitDynamics.pdf](https://github.com/idaholab/moose/files/2725276/MOOSEExplicitDynamics.pdf)
I summarized the formulation in this document. This is a first-order accurate time integrator typically used for explicit dynamics.

**Comment by hugary1995** [comment]
@bwspenc @sveerara @friedmud @cbolisetti This is ready for review.

**Comment by bwspenc** [comment]
@hugary1995 I'm talking with @cbolisetti about this right now.  My understanding of this is that your implementation of this calculates the time derivatives after the solve, which is not how we should do this. @cbolisetti has an implementation of this that computes the time derivatives before the solve, which is the way @friedmud wants us to do that. @cbolisetti thinks his code is getting pretty close, so we should probably hold off on doing anything with this until he submits a PR.

**Comment by hugary1995** [comment]
@bwspenc Yes, they talked to me and I am now aware of this issue. I am going to split the "initial condition" part of this PR out into a separate PR. I have several dues so I'll get to this the coming weekend.

File: framework/src/timeintegrators/ActuallyExplicitEuler.C [hunk 1]
```diff
@@ -96,6 +108,135 @@ ActuallyExplicitEuler::initialSetup()
 void
 ActuallyExplicitEuler::init()
 {
+  auto & es = _fe_problem.es();
+  auto & nonlinear_system = _fe_problem.getNonlinearSystemBase();
+  auto & libmesh_system = dynamic_cast<NonlinearImplicitSystem &>(nonlinear_system.system());
+  auto & system_matrix = *libmesh_system.matrix;
+
+  switch (_solve_type)
```
Comment (friedmud) [inline framework/src/timeintegrators/ActuallyExplicitEuler.C]: What is all of this doing here in init()?  It doesn't make sense... and is a copy of the stuff below.

File: framework/src/timeintegrators/ActuallyExplicitEuler.C [hunk 2]
```diff
@@ -105,85 +246,161 @@ ActuallyExplicitEuler::preSolve()
...
           es.parameters.get<Real>("linear solver tolerance"),
           es.parameters.get<unsigned int>("linear solver maximum iterations"));
 
       converged = checkLinearConvergence();
+      _n_linear_iterations = num_its_and_final_tol_EEC.first;
+
+      // Solve for explicit euler update EEU
+      // [M + C] {EEU} = {R}
+      _explicit_euler_update.zero();
+      const auto num_its_and_final_tol_EEU = _linear_solver->solve(
```
Comment (friedmud) [inline framework/src/timeintegrators/ActuallyExplicitEuler.C]: Why are there two linear solves here?  That doesn't make sense... and will be slow.

Comment (friedmud) [reply framework/src/timeintegrators/ActuallyExplicitEuler.C]: We've completely linearized the problem so I can't understand why there would be two linear solves.

**Review by friedmud — CHANGES_REQUESTED** [review]
Couple of questions to get started with...

File: framework/src/timeintegrators/ActuallyExplicitEuler.C (hunk 1 above)
Comment (hugary1995) [reply framework/src/timeintegrators/ActuallyExplicitEuler.C]: We have u_dotdot in the system, therefore we need to define a initial u_dot (though default to zero). Once we have a initial u_dot, we need to compute an initial u_dotdot from the given u_dot to get the explicit algorithm started. I talked about this part in the write-up I attached above. Even if I change the formulation, I don't think we can avoid this initialization. There are quite a lot code duplication here, and I can try to simplify it a bit.

File: framework/src/timeintegrators/ActuallyExplicitEuler.C (hunk 2 above)
Comment (hugary1995) [reply framework/src/timeintegrators/ActuallyExplicitEuler.C]: I can change it to one linear solve following @cbolisetti 's formulation.

**Comment by hugary1995** [comment]
Thanks for your comments @friedmud , I won't be able to work on this until Feb 20th. After that I can resolve the conflicts, rebase and address your comments.

**Pull request closed without merging**